src/collectors/portcheck/portcheck.py

```diff
@@ -0,0 +1,108 @@
+"""
+The PortCheckCollector checks ports listed in config file.
+
+##### Dependencies
+
+* psutil
+
+Example config file PortCheckCollector.conf
+
+```
+enabled = True
+ttl = 150
+
+[port]
+[[echo]]
+number = 8080
+
+```
+
+"""
+
+from collections import defaultdict
+import diamond.collector
+
+try:
+    import psutil
+except ImportError:
+    psutil = None
+
+try:
+    import netuitive
+except ImportError:
+    netuitive = None
+
+
+def get_port_stats(port):
+    """
+    Iterate over connections and count states for specified port
+    :param port: port for which stats are collected
+    :return: Counter with port states
+    """
+    cnts = defaultdict(int)
+    for c in psutil.net_connections('tcp'):
+        c_port = c.laddr[1]
+        if c_port != port:
+            continue
+        status = c.status.lower()
+        cnts[status] += 1
+    return cnts
+
+
+class PortCheckCollector(diamond.collector.Collector):
+
+    def __init__(self, *args, **kwargs):
+        super(PortCheckCollector, self).__init__(*args, **kwargs)
+        self.hostname = self.get_hostname()
+        self.ttl = self.config['ttl']
+
+        if not netuitive:
+            self.log.error('netuitive import failed. PortCheckCollector disabled')
+            self.enabled = False
+            return
+
+        try:
+            self.version = self._get_version()
+            self.api = netuitive.Client(self.config['url'], self.config['api_key'], self.version)
+        except Exception as e:
+            self.log.debug(e)
+
+        self.ports = {}
+        for port_name, cfg in self.config['port'].items():
+            port_cfg = {}
+            for key in ('number',):
+                port_cfg[key] = cfg.get(key, [])
+            self.ports[port_name] = port_cfg
+
+    def get_default_config_help(self):
+        config_help = super(PortCheckCollector, self).get_default_config_help()
+        config_help.update({
+        })
+        return config_help
+
+    def get_default_config(self):
+        config = super(PortCheckCollector, self).get_default_config()
+        config.update({
+            'path': 'port',
+            'port': {},
+        })
+        return config
+
+    def collect(self):
+        """
+        Overrides the Collector.collect method
+        """
+
+        if psutil is None:
+            self.log.error('Unable to import module psutil')
+            return {}
+
+        for port_name, port_cfg in self.ports.iteritems():
```

Comment:
It looks like for each check you are getting all the port stats then searching for the port.
Would it be possible to make just one call to psutil.net_connections and reuse the results?

Reply:
yeah, I inherit it from PortStatCollector. I can optimize even though I don't think there will be a lot of ports configured on each server.